src/lib/Inference.hs

```diff
@@ -269,8 +269,10 @@ instance ExtOutMap InfOutMap EnvFrag where
   extendOutMap (InfOutMap bindings ss dd oldUn) frag =
     withExtEvidence frag do
       let newUn = UnsolvedEnv $ getAtomNames frag
-      -- as an optimization, only do the zonking for the new stuff
       let newEnv = bindings `extendOutMap` frag
+      -- As an optimization, only do the zonking for the new stuff.
+      -- Also, we zonk unconditionally, even when the solver substitution
```

Comment:
Maybe this second comment ("Also, ...")  should go with the new definition of `zonkUnsolvedEnv`, since nothing has changed here.

Reply:
Good idea. I moved it there!